#!/bin/bash

# Call variants using pileups of tumor and normal

# $1 - tumor pileup file generated by Waltz

# $2 - normal pileup file generated by Waltz. The pileup files must have the same number of lines corresponding to same positions

# $3 - sample name to be used for output files

# $4 - hotspots file

# $5 - normals noise frequencies file for polishing 


export TMPDIR=/ifs/work/scratch

java="/opt/common/CentOS_6/java/jdk1.8.0_31/bin/java"

tumorPileup=$1
normalPileup=$2
sample=$3
hotspotsFile=$4
noiseFrequenciesFile=$5


echo -e "`date` Calling Variants"

# call variants
$java -server -Xms8g -Xmx8g -cp ~/software/Marianas.jar org.mskcc.marianas.variantcalling.VariantCaller $tumorPileup $normalPileup $sample $hotspotsFile $noiseFrequenciesFile



echo -e "`date` Annotating Variants"

# annotate with maf2maf (VEP)

# this may be improved further with ref allele, ref count, normal alleles etc.
# make input maf
awk 'BEGIN{FS=OFS="\t"; print "Chromosome\tStart_Position\tReference_Allele\tTumor_Seq_Allele2\tTumor_Sample_Barcode\tt_depth\tt_alt_count\tpolishing_pvalue\tn_depth\tvariant_id"}{if(NR==1) next; print $2, $3, $4, $5, $1, $9, $6, $8, $14, $19"."$20}' $sample-calls.txt > $sample-input-maf.maf

# call maf2maf
cmo_maf2maf --input-maf $sample-input-maf.maf --output-maf $sample.maf #--vep-forks 4
#rm $sample-input-maf.maf

# Not doing this right now
# bring main annotations back to the sample calls file
#awk 'BEGIN{FS=OFS="\t"}{}' $sample.maf > $sample-calls-modified.txt
#mv $sample-calls-modified.txt $sample-calls.txt


echo -e "`date` Done."


#
